ffi/src/ffi_tracing.rs

```diff
@@ -0,0 +1,555 @@
+//! FFI functions to allow engines to receive log and tracing events from kernel
+
+use std::sync::{Arc, Mutex};
+use std::{fmt, io};
+
+use delta_kernel::{DeltaResult, Error};
+use tracing::{
+    field::{Field as TracingField, Visit},
+    Event as TracingEvent, Subscriber,
+};
+use tracing_subscriber::fmt::MakeWriter;
+use tracing_subscriber::{filter::LevelFilter, layer::Context, registry::LookupSpan, Layer};
+
+use crate::{kernel_string_slice, KernelStringSlice};
+
+/// Definitions of level verbosity. Verbose Levels are "greater than" less verbose ones. So
+/// Level::ERROR is the lowest, and Level::TRACE the highest.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub enum Level {
+    ERROR = 0,
+    WARN = 1,
+    INFO = 2,
+    DEBUG = 3,
+    TRACE = 4,
+}
+
+impl Level {
+    fn is_valid(value: u32) -> bool {
+        static VALID_VALUES: &[u32] = &[0, 1, 2, 3, 4];
+        VALID_VALUES.contains(&value)
+    }
+}
+
+impl From<&tracing::Level> for Level {
```

Comment:
is `tracing::Level` not `Copy`, that we have to consume a ref instead?

Or does it just so happen all the call sites already have a reference and this avoids the need for `(*level).into()`?

Reply:
yeah, I use this only as `metadata.level().into()`, and [metadata.level()](https://docs.rs/tracing-core/0.1.33/tracing_core/metadata/struct.Metadata.html#method.level) returns `&Level`, so this makes it a little cleaner at the usage site.